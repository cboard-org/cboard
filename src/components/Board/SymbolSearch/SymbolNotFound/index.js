export { default } from './SymbolNotFound.component';
