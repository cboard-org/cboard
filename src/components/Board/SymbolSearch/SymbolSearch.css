.react-autosuggest__container {
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
}

.react-autosuggest__input {
  width: calc(100% - 16px);
  height: 32px;
  padding: 0 16px;
  border-radius: 4px;
  font-size: 1em;
  font-weight: 500;
  font-family: 'Helvetica', 'Arial', sans-serif;
  border: 0;
  color: #fff;
  background: rgba(0, 0, 0, 0.25);
  outline: 0;
}

.react-autosuggest__input::placeholder {
  color: rgba(255, 255, 255, 0.26);
}

.react-autosuggest__clear {
  margin-right: 1em;
}

.react-autosuggest__suggestions-container {
  position: absolute;
  top: 120px;
  left: 0;
  width: 100%;
  overflow: hidden;
  z-index: 1;
  color: #222;
  background: #fff;
  overflow-x: hidden;
  overflow-y: scroll;
  -webkit-overflow-scrolling: touch;
}

@media (min-width: 600px) {
  .react-autosuggest__suggestions-container {
    top: 128px;
  }
}

.react-autosuggest__suggestions-container--open {
  height: fit-content;
}

.react-autosuggest__suggestions-list {
  list-style: none;
  padding: 0;
  margin: 0;
  display: flex;
  flex-wrap: wrap;
  background: rgba(63, 81, 181, 0.2);
}

.react-autosuggest__suggestion {
  width: 25%;
  height: 100px;
  overflow-wrap: break-word;
  word-wrap: break-word;
  hyphens: auto;
}

@media (min-width: 600px) {
  .react-autosuggest__suggestion {
    width: 20%;
  }
}

.SymbolSearch__Suggestion {
  height: 100%;
  width: 100%;
  border: 2px solid transparent;
  cursor: pointer;
}

.SymbolSearch__Suggestion--highlighted {
  border-color: #9d3fb5;
}
