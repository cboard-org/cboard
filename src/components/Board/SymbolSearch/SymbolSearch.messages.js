import { defineMessages } from 'react-intl';

export default defineMessages({
  searchSymbolLibrary: {
    id: 'cboard.components.SymbolSearch.searchSymbolLibrary',
    defaultMessage: 'Search symbol library'
  },
  uploadAnImage: {
    id: 'cboard.components.InputImage.uploadImage',
    defaultMessage: 'Upload an image'
  }
});
