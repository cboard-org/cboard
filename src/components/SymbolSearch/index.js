export { default } from './SymbolSearch.component';
