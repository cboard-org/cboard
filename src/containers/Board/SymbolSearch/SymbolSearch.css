.react-autosuggest__suggestions-list {
  list-style: none;
  padding: 0;
  margin: 0;
  display: flex;
  flex-wrap: wrap;
  background: rgb(255,255,255); /* The Fallback */
  background: rgba(63, 81, 181, 0.2);
}

.react-autosuggest__suggestion {
  width: 25%;
  overflow-wrap: break-word;
  word-wrap: break-word;
  hyphens: auto;
}

@media (min-width: 600px) {
  .react-autosuggest__suggestion {
    width: 12.5%;
  }
}

.SymbolSearch__Suggestion--highlighted {
  border: 2px solid #3f51b5;
  padding: 2px;
}

.SymbolSearch__Suggestion-img {
  width: 100%;
  height: auto;
}
