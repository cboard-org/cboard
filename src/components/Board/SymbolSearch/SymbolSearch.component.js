import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import Autosuggest from 'react-autosuggest';
import classNames from 'classnames';
import isMobile from 'ismobilejs';
import queryString from 'query-string';

import FullScreenDialog from '../../UI/FullScreenDialog';
import Symbol from '../Symbol';
import messages from './SymbolSearch.messages';
import './SymbolSearch.css';
import API from '../../../api';
import {
  ARASAAC_BASE_PATH_API,
  TAWASOL_BASE_IMAGE_ULR
} from '../../../constants';

export class SymbolSearch extends PureComponent {
  static propTypes = {
    intl: intlShape.isRequired,
    open: PropTypes.bool,
    maxSuggestions: PropTypes.number,
    onChange: PropTypes.func.isRequired,
    onClose: PropTypes.func.isRequired
  };

  static defaultProps = {
    open: false,
    maxSuggestions: 16
  };

  state = {
    value: '',
    suggestions: [],
    skin: undefined,
    hair: undefined
  };

  symbols = [];

  async componentDidMount() {
    const {
      intl: { locale }
    } = this.props;
    try {
      const languagesResponse = await API.getLanguage(`${locale}-`);
      const { skin, hair } = languagesResponse;
      if (skin && hair) await this.setState({ skin, hair });
    } catch (err) {}

    import('../../../api/mulberry-symbols.json').then(
      ({ default: mulberrySymbols }) => {
        this.symbols = this.translateSymbols(mulberrySymbols);
      }
    );
  }

  translateSymbols(symbols = []) {
    return symbols.map(symbol => {
      const translatedId = this.props.intl
        .formatMessage({ id: symbol.id })
        .replace(/[\u0591-\u05C7]/g, '') // todo: not on every locale - strip hebrew niqqud
        .toLowerCase();

      return { ...symbol, translatedId };
    });
  }

  getSuggestionValue(suggestion) {
    return suggestion.id;
  }

  getSuggestions(value) {
    const { maxSuggestions } = this.props;
    const inputValue = value.trim().toLowerCase();
    const inputLength = inputValue.length;
    let count = 0;

    return this.symbols.filter(symbol => {
      if (count >= maxSuggestions) {
        return false;
      }
      const translatedId = symbol.translatedId;
      let keep = translatedId.slice(0, inputLength) === inputValue;

      if (!keep) {
        const words = translatedId.split(' ');

        for (let i = 1; i < words.length; i += 1) {
          keep = words[i].slice(0, inputLength) === inputValue;

          if (keep) {
            break;
          }
        }
      }

      if (keep) {
        count += 1;
      }
      return keep;
    });
  }

  fetchSrasaacSuggestions = async searchText => {
    const {
      intl: { locale }
    } = this.props;
    const { skin, hair } = this.state;
    try {
      const data = await API.arasaacPictogramsSearch(locale, searchText);
      if (data.length) {
        const suggestions = [
          ...this.state.suggestions.filter(
            suggestion => !suggestion.fromArasaac
          )
        ];
        const arasaacSuggestions = data.map(
          ({ idPictogram, keywords: [keyword] }) => {
            return {
              id: keyword.keyword,
              src: `${ARASAAC_BASE_PATH_API}pictograms/${idPictogram}?${queryString.stringify(
                { skin, hair }
              )}`,
              translatedId: keyword.keyword,
              fromArasaac: true
            };
          }
        );
        this.setState({ suggestions: [...suggestions, ...arasaacSuggestions] });
      }
      return [];
    } catch (err) {
      return [];
    }
  };

  fetchTawasolSuggestions = async searchText => {
    const {
      intl: { locale }
    } = this.props;

    try {
      const data = await API.tawasolPictogramsSearch(locale, searchText);
      if (data.length) {
        const suggestions = [
          ...this.state.suggestions.filter(
            suggestion => !suggestion.fromTawasol
          )
        ];
        const tawasolSuggestions = data
          .filter(pictogram => pictogram.source_id === '1')
          .map(({ description, image_uri }) => {
            return {
              id: description,
              src: `${TAWASOL_BASE_IMAGE_ULR}${image_uri}`,
              translatedId: description,
              fromTawasol: true
            };
          });
        this.setState({ suggestions: [...suggestions, ...tawasolSuggestions] });
      }
      return [];
    } catch (err) {
      return [];
    }
  };

  handleSuggestionsFetchRequested = async ({ value }) => {
    const arabic = /[\u0600-\u06FF]/;
    let localSuggestions;

    if (arabic.test(value)) {
      this.fetchTawasolSuggestions(value);
    }
    localSuggestions = this.getSuggestions(value);
    this.fetchSrasaacSuggestions(value);

    // Tawasol's suggestions may include some non-arabic strings
    this.setState({
      suggestions: [...localSuggestions]
    });
  };

  handleSuggestionsClearRequested = () => {
    this.setState({
      suggestions: []
    });
  };

  handleSuggestionSelected = (event, { suggestion }) => {
    const { onChange, onClose } = this.props;
    this.setState({ value: '' });

    onChange({
      image: suggestion.src,
      label: suggestion.translatedId,
      labelKey: suggestion.id
    });
    onClose();
  };

  handleChange = (event, { newValue }) => {
    this.setState({
      value: newValue
    });
  };

  renderSuggestion(suggestion, { query, isHighlighted }) {
    const suggestionClassName = classNames('SymbolSearch__Suggestion', {
      'SymbolSearch__Suggestion--highlighted': isHighlighted
    });

    return (
      <div className={suggestionClassName}>
        <Symbol label={suggestion.translatedId} image={suggestion.src} />
      </div>
    );
  }

  renderSuggestionsContainer(options) {
    const { containerProps, children } = options;
    return <div {...containerProps}>{children}</div>;
  }

  render() {
    const { intl, open, onClose } = this.props;

    const autoSuggest = (
      <Autosuggest
        aria-label="Search auto-suggest"
        suggestions={this.state.suggestions}
        focusInputOnSuggestionClick={!isMobile.any}
        onSuggestionsFetchRequested={this.handleSuggestionsFetchRequested}
        onSuggestionsClearRequested={this.handleSuggestionsClearRequested}
        onSuggestionSelected={this.handleSuggestionSelected}
        renderSuggestionsContainer={this.renderSuggestionsContainer}
        getSuggestionValue={this.getSuggestionValue}
        renderSuggestion={this.renderSuggestion}
        highlightFirstSuggestion={true}
        inputProps={{
          autoFocus: true,
          placeholder: intl.formatMessage(messages.searchSymbolLibrary),
          label: intl.formatMessage(messages.searchSymbolLibrary),
          value: this.state.value,
          onChange: this.handleChange
        }}
      />
    );

    return (
      <FullScreenDialog
        open={open}
        buttons={autoSuggest}
        transition="fade"
        onClose={onClose}
      />
    );
  }
}

export default injectIntl(SymbolSearch);
