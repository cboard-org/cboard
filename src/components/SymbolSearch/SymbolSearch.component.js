import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import Autosuggest from 'react-autosuggest';
import classNames from 'classnames';
import isMobile from 'ismobilejs';

import messages from './SymbolSearch.messages';
import mulberrySymbols from '../../api/mulberry-symbols.json';
import FullScreenDialog from '../FullScreenDialog';
import Symbol from '../Symbol';
import './SymbolSearch.css';

export class SymbolSearchComponent extends PureComponent {
  static propTypes = {
    intl: intlShape.isRequired,
    open: PropTypes.bool,
    maxSuggestions: PropTypes.number,
    onChange: PropTypes.func.isRequired,
    onRequestClose: PropTypes.func.isRequired
  };

  static defaultProps = {
    open: false,
    maxSuggestions: 16
  };

  state = {
    value: '',
    suggestions: []
  };

  symbols = this.translateSymbols(mulberrySymbols);

  translateSymbols(symbols) {
    return symbols.map(symbol => {
      const translatedId = this.props.intl
        .formatMessage({ id: symbol.id })
        .replace(/[\u0591-\u05C7]/g, '') // todo: not on every locale - strip hebrew niqqud
        .toLowerCase();

      return { ...symbol, translatedId };
    });
  }

  getSuggestionValue(suggestion) {
    return suggestion.id;
  }

  getSuggestions(value) {
    const { maxSuggestions } = this.props;
    const inputValue = value.trim().toLowerCase();
    const inputLength = inputValue.length;
    let count = 0;

    return this.symbols.filter(symbol => {
      if (count >= maxSuggestions) {
        return false;
      }
      const translatedId = symbol.translatedId;
      let keep = translatedId.slice(0, inputLength) === inputValue;

      if (!keep) {
        const words = translatedId.split(' ');

        for (let i = 1; i < words.length; i += 1) {
          keep = words[i].slice(0, inputLength) === inputValue;

          if (keep) {
            break;
          }
        }
      }

      if (keep) {
        count += 1;
      }
      return keep;
    });
  }

  handleSuggestionsFetchRequested = ({ value }) => {
    this.setState({
      suggestions: this.getSuggestions(value)
    });
  };

  handleSuggestionsClearRequested = () => {
    this.setState({
      suggestions: []
    });
  };

  handleSuggestionSelected = (event, { suggestion }) => {
    const { onChange, onRequestClose } = this.props;
    this.setState({ value: '' });

    onChange({
      img: suggestion.src,
      labelKey: suggestion.id
    });
    onRequestClose();
  };

  handleChange = (event, { newValue }) => {
    this.setState({
      value: newValue
    });
  };

  renderSuggestion(suggestion, { query, isHighlighted }) {
    const suggestionClassName = classNames('SymbolSearch__Suggestion', {
      'SymbolSearch__Suggestion--highlighted': isHighlighted
    });

    return (
      <div className={suggestionClassName}>
        <Symbol label={suggestion.translatedId} img={suggestion.src} />
      </div>
    );
  }

  renderSuggestionsContainer(options) {
    const { containerProps, children } = options;
    return <div {...containerProps}>{children}</div>;
  }

  render() {
    const { intl, open, onRequestClose } = this.props;

    const autoSuggest = (
      <Autosuggest
        aria-label="Search auto-suggest"
        suggestions={this.state.suggestions}
        focusInputOnSuggestionClick={!isMobile.any}
        onSuggestionsFetchRequested={this.handleSuggestionsFetchRequested}
        onSuggestionsClearRequested={this.handleSuggestionsClearRequested}
        onSuggestionSelected={this.handleSuggestionSelected}
        renderSuggestionsContainer={this.renderSuggestionsContainer}
        getSuggestionValue={this.getSuggestionValue}
        renderSuggestion={this.renderSuggestion}
        highlightFirstSuggestion={true}
        inputProps={{
          autoFocus: true,
          placeholder: intl.formatMessage(messages.searchImageLibrary),
          label: intl.formatMessage(messages.searchImageLibrary),
          value: this.state.value,
          onChange: this.handleChange
        }}
      />
    );

    return (
      <FullScreenDialog
        open={open}
        buttons={autoSuggest}
        transition="fade"
        onRequestClose={onRequestClose}
      />
    );
  }
}

export default injectIntl(SymbolSearchComponent);
