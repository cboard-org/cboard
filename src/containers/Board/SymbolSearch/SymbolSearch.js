import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import Autosuggest from 'react-autosuggest';
import TextField from 'material-ui/TextField';
import classNames from 'classnames';

import messages from './messages';
import mulberrySymbols from '../../../api/mulberry-symbols.json';
import './SymbolSearch.css';

export class SymbolSearch extends Component {
  state = {
    value: '',
    suggestions: []
  };

  symbols = this.translateSymbols(mulberrySymbols);

  translateSymbols(symbols) {
    return symbols.map(symbol => {
      const translatedId = this.props.intl
        .formatMessage({ id: symbol.id })
        .replace(/[\u0591-\u05C7]/g, '') // todo: not on every locale - strip hebrew niqqud
        .toLowerCase();

      return { ...symbol, translatedId };
    });
  }

  getSuggestionValue(suggestion) {
    return suggestion.id;
  }

  getSuggestions(value) {
    const { maxSuggestions } = this.props;
    const inputValue = value.trim().toLowerCase();
    const inputLength = inputValue.length;
    let count = 0;

    return this.symbols.filter(symbol => {
      if (count >= maxSuggestions) {
        return false;
      }
      const translatedId = symbol.translatedId;
      let keep = translatedId.slice(0, inputLength) === inputValue;

      if (!keep) {
        const words = translatedId.split(' ');

        for (let i = 1; i < words.length; i += 1) {
          keep = words[i].slice(0, inputLength) === inputValue;

          if (keep) {
            break;
          }
        }
      }

      if (keep) {
        count += 1;
      }
      return keep;
    });
  }

  handleSuggestionsFetchRequested = ({ value }) => {
    this.setState({
      suggestions: this.getSuggestions(value)
    });
  };

  handleSuggestionsClearRequested = () => {
    this.setState({
      suggestions: []
    });
  };

  handleSuggestionSelected = (event, { suggestion }) => {
    const { onChange } = this.props;
    this.setState({ value: '' });

    onChange({
      img: suggestion.src,
      label: suggestion.id
    });
  };

  handleChange = (event, { newValue }) => {
    this.setState({
      value: newValue
    });
  };

  renderInput(inputProps) {
    const { value, label, ref, ...other } = inputProps;

    return (
      <TextField
        fullWidth
        value={value}
        label={label}
        inputRef={ref}
        InputProps={{
          ...other
        }}
      />
    );
  }

  renderSuggestion(suggestion, { query, isHighlighted }) {
    const suggestionClassName = classNames({
      SymbolSearch__Suggestion: true,
      'SymbolSearch__Suggestion--highlighted': isHighlighted
    });

    return (
      <div className={suggestionClassName}>
        <img
          className="SymbolSearch__Suggestion-img"
          src={suggestion.src}
          alt=""
        />
        <div>{suggestion.translatedId}</div>
      </div>
    );
  }

  renderSuggestionsContainer(options) {
    const { containerProps, children } = options;
    return <div {...containerProps}>{children}</div>;
  }

  render() {
    const { intl } = this.props;
    return (
      <Autosuggest
        renderInputComponent={this.renderInput}
        suggestions={this.state.suggestions}
        onSuggestionsFetchRequested={this.handleSuggestionsFetchRequested}
        onSuggestionsClearRequested={this.handleSuggestionsClearRequested}
        onSuggestionSelected={this.handleSuggestionSelected}
        renderSuggestionsContainer={this.renderSuggestionsContainer}
        getSuggestionValue={this.getSuggestionValue}
        renderSuggestion={this.renderSuggestion}
        highlightFirstSuggestion={true}
        inputProps={{
          autoFocus: true,
          placeholder: intl.formatMessage(messages.searchImageLibrary),
          label: intl.formatMessage(messages.searchImageLibrary),
          value: this.state.value,
          onChange: this.handleChange
        }}
      />
    );
  }
}

SymbolSearch.propTypes = {
  intl: intlShape.isRequired,
  maxSuggestions: PropTypes.number,
  onChange: PropTypes.func.isRequired
};

SymbolSearch.defaultProps = {
  maxSuggestions: 16
};

export default injectIntl(SymbolSearch);
