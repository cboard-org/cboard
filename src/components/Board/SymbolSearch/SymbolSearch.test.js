import React from 'react';
import { shallowMatchSnapshot } from '../../../common/test_utils';
import SymbolSearch from './SymbolSearch.component';

// import { IntlProvider } from 'react-intl';

describe('SymbolSearch tests', () => {
  test('default renderer', () => {
    shallowMatchSnapshot(
      // <IntlProvider locale="en">
      <SymbolSearch />
      // </IntlProvider>
    );
  });
});
