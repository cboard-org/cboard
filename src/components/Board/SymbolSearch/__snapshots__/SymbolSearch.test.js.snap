// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SymbolSearch tests default renderer 1`] = `<InjectIntl(SymbolSearch) />`;
